table/two_level_iterator.cc

```diff
@@ -22,22 +22,9 @@ namespace {
 class TwoLevelIterator : public InternalIterator {
  public:
   explicit TwoLevelIterator(TwoLevelIteratorState* state,
-                            InternalIterator* first_level_iter,
-                            bool need_free_iter_and_state);
+                            InternalIterator* first_level_iter);
 
-  virtual ~TwoLevelIterator() {
-    // Assert that the TwoLevelIterator is never deleted while Pinning is
-    // Enabled.
-    assert(!pinned_iters_mgr_ ||
-           (pinned_iters_mgr_ && !pinned_iters_mgr_->PinningEnabled()));
-    first_level_iter_.DeleteIter(!need_free_iter_and_state_);
-    second_level_iter_.DeleteIter(false);
-    if (need_free_iter_and_state_) {
-      delete state_;
-    } else {
-      state_->~TwoLevelIteratorState();
-    }
-  }
+  virtual ~TwoLevelIterator() { delete state_; }
```

Comment:
Under the new design, who's responsible for deleting first_level_iter_ and second_level_iter_?
RIght now there are memory leaks that trace back to https://github.com/facebook/rocksdb/blob/master/table/two_level_iterator.cc#L205 and https://github.com/facebook/rocksdb/blob/master/table/two_level_iterator.cc#L196

Call stack see https://gist.github.com/miasantreble/0de9bd7ee11451cb1bebbd108f7b3e80

Reply:
@miasantreble this seems like a bug to me.